### PR TITLE
ESQL: Test utilities for functions

### PR DESCRIPTION
**Pull request by nik9000**
Adds utilities to generate test cases for `null` arguments and combinations of arguments that aren't supported. The `null` arguments cases are asserted to return `null` which is how *most* functions work. The invalid argument type cominations are asserted to return an error.

This slightly modifies some functions to explicitly return `null` if given an always-null argument. It's simpler to test this way and easier to line up. So doing something like `GREATEST(1, 2, null)` will both return `null` and have the type `null`.

### CODE REVIEW

File: x-pack/plugin/esql/compute/src/main/java/org/elasticsearch/compute/operator/EvalOperator.java [hunk 1]
```diff
@@ -51,5 +51,15 @@ public interface ExpressionEvaluator {
...
+        @Override
+        public Block eval(Page page) {
+            return Block.constantNullBlock(page.getPositionCount());
+        }
+
+        @Override
+        public String toString() {
+            return "ConstantNull";
+        }
+    };
```
Comment (nik9000) [inline x-pack/plugin/esql/compute/src/main/java/org/elasticsearch/compute/operator/EvalOperator.java:64]: This `toString` shows up in explain output so it's nice not to get a random closure.

File: x-pack/plugin/esql/src/test/java/org/elasticsearch/xpack/esql/expression/function/scalar/AbstractScalarFunctionTestCase.java [hunk 2]
```diff
@@ -103,12 +106,6 @@ protected final DataType[] representable() {
 
     protected record ArgumentSpec(boolean optional, Set<DataType> validTypes) {}
 
-    public final void testSimpleResolveTypeValid() {
-        // TODO: The expected output type should probably be on the TestCase
-        Expression expression = buildFieldExpression(testCase);
-        assertResolveTypeValid(expression, expectedType(expression.children().stream().map(e -> e.dataType()).toList()));
-    }
-
```
Comment (nik9000) [inline x-pack/plugin/esql/src/test/java/org/elasticsearch/xpack/esql/expression/function/scalar/AbstractScalarFunctionTestCase.java:111]: This is well covered by the test cases we build now.

File: x-pack/plugin/esql/src/test/java/org/elasticsearch/xpack/esql/expression/function/scalar/math/AcosTests.java [hunk 3]
```diff
@@ -10,43 +10,22 @@
...
-
-    @Override
-    protected DataType expectedType(List<DataType> argTypes) {
-        return DataTypes.DOUBLE;
-    }
-
-    @Override
-    protected List<ArgumentSpec> argSpec() {
-        return List.of(required(numerics()));
+        List<TestCaseSupplier> suppliers = TestCaseSupplier.forUnaryCastingToDouble("AcosEvaluator", "val", Math::acos);
```
Comment (nik9000) [inline x-pack/plugin/esql/src/test/java/org/elasticsearch/xpack/esql/expression/function/scalar/math/AcosTests.java:27]: This generates a dozen or so case using the `Math:acos` oracle. If we need hand rolled cases we can explicitly add them to the `suppliers` list here.

**Review by nik9000 — COMMENTED** [review]
I've reverted the null handling changes because that's handled by an optimizer rule. I run that rule now in the tests so I can *mostly* keep the null assertions.

File: x-pack/plugin/esql/src/test/java/org/elasticsearch/xpack/esql/expression/function/AbstractFunctionTestCase.java [hunk 4]
```diff
@@ -340,4 +537,210 @@ public final void testSimpleConstantFolding() {
...
+            .collect(Collectors.toSet())
+            .stream()
+            .flatMap(count -> allPermutations(count))
+            .filter(types -> valid.contains(types) == false)
+            /*
+             * Skip any cases with more than one null. Our tests don't generate
+             * the full combinatorial explosions of all nulls - just a single null.
+             * Hopefully <null>, <null> cases will function the same as <null>, <valid>
+             * cases.
+             */
```
Comment (alex-spies) [inline x-pack/plugin/esql/src/test/java/org/elasticsearch/xpack/esql/expression/function/AbstractFunctionTestCase.java:645]: nit: the fact that we don't generate test cases with more than 1 null would probably also be great to know in the doc string of `errorsForCasesWithoutExamples`.

File: x-pack/plugin/esql/src/test/java/org/elasticsearch/xpack/esql/expression/function/AbstractFunctionTestCase.java [hunk 5]
```diff
@@ -340,4 +537,210 @@ public final void testSimpleConstantFolding() {
     public void testSerializationOfSimple() {
         assertSerialization(buildFieldExpression(testCase));
     }
+
+    /**
+     * Adds cases with {@code null} and asserts that the result is {@code null}.
+     */
+    protected static List<TestCaseSupplier> anyNullIsNull(boolean entirelyNullPreservesType, List<TestCaseSupplier> testCaseSuppliers) {
```
Comment (alex-spies) [inline x-pack/plugin/esql/src/test/java/org/elasticsearch/xpack/esql/expression/function/AbstractFunctionTestCase.java:544]: nit: would be cool if the doc string explained the params a bit, esp. `entirelyNullPreservesType`.

**Review by alex-spies — APPROVED** [review]
Looking good to me. I like that this adds a lot of negative tests without really requiring any additional work to be done in each individual test class.

File: x-pack/plugin/esql/src/test/java/org/elasticsearch/xpack/esql/expression/function/AbstractFunctionTestCase.java [hunk 6]
```diff
@@ -138,37 +154,198 @@ public Matcher<Object> getMatcher() {
...
+                }
+            }
+            return suppliers;
+        }
+
+        private static final Map<DataType, List<Map.Entry<String, Supplier<Object>>>> RANDOM_VALUE_SUPPLIERS = Map.ofEntries(
+            Map.entry(
+                DataTypes.DOUBLE,
+                List.of(
+                    Map.entry("<0 double>", () -> 0.0d),
```
[the remark this answers is not in the document]
Comment (nik9000) [reply x-pack/plugin/esql/src/test/java/org/elasticsearch/xpack/esql/expression/function/AbstractFunctionTestCase.java:298]: @not-napoleon has some opinions here, but let's get something in for now. He's working on nan-things. Maybe we still want -0, positive infinity, and negative infinity though.

File: x-pack/plugin/esql/src/test/java/org/elasticsearch/xpack/esql/expression/function/AbstractFunctionTestCase.java (hunk 4 above)
Comment (nik9000) [reply x-pack/plugin/esql/src/test/java/org/elasticsearch/xpack/esql/expression/function/AbstractFunctionTestCase.java:645]: ++

File: x-pack/plugin/esql/src/test/java/org/elasticsearch/xpack/esql/expression/function/AbstractFunctionTestCase.java (hunk 5 above)
Comment (nik9000) [reply x-pack/plugin/esql/src/test/java/org/elasticsearch/xpack/esql/expression/function/AbstractFunctionTestCase.java:544]: ++. That was one I had to add later on to make everything make sense and I didn't doc it properly. Thanks for pointing it out!

**Pull request merged**